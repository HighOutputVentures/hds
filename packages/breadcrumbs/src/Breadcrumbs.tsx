import {
  As,
  Breadcrumb,
  BreadcrumbItem,
  BreadcrumbLink,
  Icon,
  Skeleton,
  SystemStyleObject,
  VisuallyHidden,
} from '@chakra-ui/react';
import * as React from 'react';
import { v4 as uuid } from 'uuid';
import ChevronRightIcon from './icons/ChevronRightIcon';
import HomeIcon from './icons/HomeIcon';

type Item = {
  href: string;
  label: string | JSX.Element;
  isActive?: boolean;
  isLoading?: boolean;
  isDisabled?: boolean;
};

export type BreadcrumbProps = {
  as?: As;
  items?: Item[];
  homeHref?: string;
  separator?: string | JSX.Element;
  withAccent?: boolean;
};

export default React.forwardRef<HTMLDivElement, BreadcrumbProps & SystemStyleObject>(
  function HdsBreadcrumbs(props, ref) {
    const {
      as,
      items,
      homeHref,
      withAccent = true,
      separator = (
        <Icon
          as={ChevronRightIcon}
          width={5}
          height={5}
          color="neutrals.300"
          display="flex"
          alignItems="center"
          justifyContent="center"
        />
      ),
      ...others
    } = props;

    const styles = useStyles({ withAccent });
    const shouldShowHome = !!homeHref;
    const shouldSelectHome = !items?.some((o) => !!o.isActive);

    return (
      <Breadcrumb
        ref={ref}
        spacing="14px"
        fontSize="14px"
        lineHeight="14px"
        separator={separator}
        data-testid="hds.breadcrumb"
        sx={others}
      >
        {shouldShowHome && (
          <BreadcrumbItem>
            <BreadcrumbLink
              as={as}
              href={homeHref}
              isCurrentPage={shouldSelectHome}
              sx={styles.link}
            >
              <Icon as={HomeIcon} w={4} h={4} />
              <VisuallyHidden>Go to Homepage</VisuallyHidden>
            </BreadcrumbLink>
          </BreadcrumbItem>
        )}

        {items?.map(({ label, href, isActive, isDisabled, isLoading }) => (
          <BreadcrumbItem key={uuid()}>
            <BreadcrumbLink
              as={as}
              href={href}
              isCurrentPage={isActive}
              onClick={(e) => {
                if (isLoading || isDisabled) {
                  e.preventDefault();
                }
              }}
              {...(isDisabled && {
                'data-disabled': true,
              })}
              {...(isLoading && {
                'data-loading': true,
              })}
              {...(isActive && {
                'data-active': true,
              })}
              sx={styles.link}
            >
              {isLoading ? <Skeleton w="64px" h="10px" /> : label}
            </BreadcrumbLink>
          </BreadcrumbItem>
        ))}
      </Breadcrumb>
    );
  },
);

function useStyles({ withAccent }: { withAccent?: boolean } = {}) {
  return {
    link: {
      color: 'neutrals.600',
      letterSpacing: '0.02em',
      textDecoration: 'none',
      _hover: {
        color: 'neutrals.800',
      },
      _disabled: {
        color: 'neutrals.600',
        cursor: 'not-allowed',
        _hover: {},
      },
      _loading: {
        color: 'neutrals.600',
        cursor: 'not-allowed',
        _hover: {},
      },
      _active: {
        color: withAccent ? 'brand.primary.900' : 'neutrals.900',
      },
    },
  };
}
