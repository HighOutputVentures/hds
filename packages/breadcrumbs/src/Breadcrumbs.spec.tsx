import { ChakraProvider } from '@chakra-ui/react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import Breadcrumbs from './Breadcrumbs';

describe('Breadcrumbs Component', () => {
  beforeEach(() => {
    render(
      <ChakraProvider>
        <Breadcrumbs
          withAccent
          items={[
            {
              href: '/settings',
              label: 'Settings',
            },
            {
              href: '#',
              label: '...',
              isDisabled: true,
            },
            {
              href: '/settings/teams',
              label: 'Teams',
              isActive: true,
            },
          ]}
        />
      </ChakraProvider>,
    );
  });

  it('should render breadcrumb', async () => {
    const breadcrumb = screen.getByTestId('hds.breadcrumb');
    expect(breadcrumb).toBeInTheDocument();
  });

  it('Should match snapshot', () => {
    const component = renderer.create(
      <ChakraProvider>
        <Breadcrumbs
          withAccent
          items={[
            {
              href: '/settings',
              label: 'Settings',
            },
            {
              href: '#',
              label: '...',
              isDisabled: true,
            },
            {
              href: '/settings/teams',
              label: 'Teams',
              isActive: true,
            },
          ]}
        />
      </ChakraProvider>,
    );

    const snapshot = component.toJSON();
    expect(snapshot).toMatchSnapshot();
    component.unmount();
  });
});
