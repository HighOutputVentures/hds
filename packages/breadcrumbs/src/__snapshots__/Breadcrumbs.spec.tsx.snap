// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs Component Should match snapshot 1`] = `
[
  <nav
    aria-label="breadcrumb"
    className="chakra-breadcrumb css-1wezd10"
    data-testid="hds.breadcrumb"
  >
    <ol
      className="chakra-breadcrumb__list css-70qvj9"
    >
      <li
        className="chakra-breadcrumb__list-item css-18biwo"
      >
        <a
          className="chakra-breadcrumb__link css-bai5w2"
          href="/settings"
          onClick={[Function]}
        >
          Settings
        </a>
        <span
          className="css-1xgwh2j"
          role="presentation"
        >
          <svg
            className="chakra-icon css-1im13y"
            fill="none"
            focusable={false}
            height="17"
            viewBox="0 0 17 17"
            width="17"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.66675 12.3333L10.6667 8.33331L6.66675 4.33331"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="1.33333"
            />
          </svg>
        </span>
      </li>
      <li
        className="chakra-breadcrumb__list-item css-18biwo"
      >
        <a
          className="chakra-breadcrumb__link css-bai5w2"
          data-disabled={true}
          href="#"
          onClick={[Function]}
        >
          ...
        </a>
        <span
          className="css-1xgwh2j"
          role="presentation"
        >
          <svg
            className="chakra-icon css-1im13y"
            fill="none"
            focusable={false}
            height="17"
            viewBox="0 0 17 17"
            width="17"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.66675 12.3333L10.6667 8.33331L6.66675 4.33331"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="1.33333"
            />
          </svg>
        </span>
      </li>
      <li
        className="chakra-breadcrumb__list-item css-18biwo"
      >
        <a
          className="chakra-breadcrumb__link css-bai5w2"
          data-active={true}
          href="/settings/teams"
          onClick={[Function]}
        >
          Teams
        </a>
      </li>
    </ol>
  </nav>,
  <span />,
  <span
    hidden={true}
    id="__chakra_env"
  />,
]
`;
